### PR TITLE
Move Nyancats into a separate mod

### PR DESCRIPTION
**Pull request by rubenwardy**
Nyancats are independent in the default mod. Nothing else uses them or their code.
Separating it into a separate mod makes it easier for subgames to remove them.
It also makes it easier for a mod to depend on nyancats, as lots of subgames
don't have them

### CODE REVIEW

**Comment by paramat** [comment]
Seems reasonable at first consideration.

**Comment by paramat** [comment]
See https://github.com/minetest/minetest/issues/4164
Hmmmm and i decided it is the responsibility of mods to clear unwanted biome/decoration/ore registrations if the mod changes the mapgen.
So we need a way to disable nyancat generation in a mod that is run after nyancat mod.
I'm thinking move ```minetest.register_on_generated(nyancat.generate)``` into a global function so that it can be redefined as an empty function later in  another mod. Then call that global function from within ```if minetest.get_mapgen_setting("mg_name") ~= "singlenode" then```

**Comment by t4im** [comment]
Not sure if this is cleaner, but an alternative might be monkey-patching the clear-functions and deactivate itself on clear as well. Let's call it nyancat-patching

```lua
local clear = minetest.clear_registered_decorations()
minetest.clear_registered_decorations = function(...)
  kill_all_nyancats()
  return clear(...)
end
```

**Comment by paramat** [comment]
^ Not keen on that.

**Comment by paramat** [comment]
Because of https://github.com/minetest/minetest/issues/4164 there's no point in having:
```if minetest.get_mapgen_setting("mg_name") ~= "singlenode" then```
as that will always fail to keep nyancats out of singlenode.
Also sometimes it may be okay to have them in singlenode, so i suggest we enable them in all mapgens.

**Comment by paramat** [comment]
```
function nyancat.on_generated()
	minetest.register_on_generated(nyancat.generate)
end
```

**Comment by paramat** [comment]
Or, since the 'generate nyancats' function is not useful on it's own, how about:
```
function nyancat.generate_nyancats()
minetest.register_on_generated(minp, maxp, seed)
	local height_min = -31000
	local height_max = -32
	if maxp.y < height_min or minp.y > height_max then
		return
	end
	local y_min = math.max(minp.y, height_min)
	local y_max = math.min(maxp.y, height_max)
	local volume = (maxp.x - minp.x + 1) * (y_max - y_min + 1) * (maxp.z - minp.z + 1)
	local pr = PseudoRandom(seed + 9324342)
	local max_num_nyancats = math.floor(volume / (16 * 16 * 16))
	for i = 1, max_num_nyancats do
		if pr:next(0, 1000) == 0 then
			local x0 = pr:next(minp.x, maxp.x)
			local y0 = pr:next(minp.y, maxp.y)
			local z0 = pr:next(minp.z, maxp.z)
			local p0 = {x = x0, y = y0, z = z0}
			nyancat.make(p0, pr:next(0, 3), pr:next(3, 15))
		end
	end
end
end

nyancat.generate_nyancats()
```

File: mods/nyancat/init.lua [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			local y0 = pr:next(minp.y, maxp.y)
+			local z0 = pr:next(minp.z, maxp.z)
+			local p0 = {x = x0, y = y0, z = z0}
+			nyancat.make(p0, pr:next(0, 3), pr:next(3, 15))
+		end
+	end
+end
+
+if minetest.get_mapgen_setting("mg_name") ~= "singlenode" then
+	minetest.register_on_generated(nyancat.generate)
```
Comment (paramat) [inline mods/nyancat/init.lua]: Should be ```nyancat.generate_nyancats``` ?

Comment (rubenwardy) [inline mods/nyancat/init.lua]: I meant to rename it to `generate`

Comment (paramat) [inline mods/nyancat/init.lua]: Makes sense, 'nyancat' is redundant here.

**Comment by paramat** [comment]
+1 with https://github.com/minetest/minetest_game/pull/1190#issuecomment-231827129 and fixes

**Comment by rubenwardy** [comment]
Done

**Comment by paramat** [comment]
:+1:

File: mods/nyancat/init.lua [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			local x0 = pr:next(minp.x, maxp.x)
+			local y0 = pr:next(minp.y, maxp.y)
+			local z0 = pr:next(minp.z, maxp.z)
+			local p0 = {x = x0, y = y0, z = z0}
+			nyancat.make(p0, pr:next(0, 3), pr:next(3, 15))
+		end
+	end
+end
+
+if minetest.get_mapgen_setting("mg_name") ~= "singlenode" then
```
Comment (paramat) [inline mods/nyancat/init.lua]: As suggested in https://github.com/minetest/minetest_game/pull/1190#issuecomment-231827129 please can this line be removed? As it has no effect anyway and nyancats will often be wanted in singlenode.

**Comment by paramat** [comment]
Added line comment.

**Comment by rubenwardy** [comment]
done

File: mods/nyancat/init.lua [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	type = "fuel",
+	recipe = "nyancat:nyancat_rainbow",
+	burntime = 1,
+})
+
+nyancat = {}
+
+function nyancat.place(pos, facedir, length)
+	local tailvec = {x = 0, y = 0, z = 0}
+	if facedir == 0 then
```
Comment (SmallJoker) [inline mods/nyancat/init.lua]: The perfect situation to use `minetest.facedir_to_dir(facedir)` here.

Comment (rubenwardy) [inline mods/nyancat/init.lua]: do you mean 

```lua
local tailvec = minetest.facedir_to_dir(facedir)
```

then remove the long if chain?

That's a fairly unrelated change

Comment (SmallJoker) [inline mods/nyancat/init.lua]: Rather 
```lua
if facedir > 3 then
    facedir = 0
end
local tailvec = minetest.facedir_to_dir(facedir)```
But yes, it shortens the code. Should be perceived as a "tweak on the fly".

**Comment by paramat** [comment]
:+1: 
(Hmmmm approves too)

**Comment by rubenwardy** [comment]
Done. Confirmed working

**Comment by paramat** [comment]
:+1:

**Comment by rubenwardy** [comment]
updated docs to document nyancat.generate

**Comment by paramat** [comment]
Nice.

File: game_api.txt [hunk 4]
```diff
@@ -255,7 +255,12 @@ Nyancat API
 
 `nyancat.place(pos, facedir, length)`
 
-Place a cat at `pos` facing `facedir` with tail length `length`
+^ Place a cat at `pos` facing `facedir` with tail length `length`
+
+`nyancat.generate(minp, maxp, seed)`
+
+^ Called by `minetest.register_on_generated`. To disable nyancat generation,
+  you can replace minetest.register_on_generated with an empty function
```
Comment (paramat) [inline game_api.txt]: Should be something like:
```To disable nyancat generation you can redefine nyancat.generate() to be an empty function.```

**Comment by paramat** [comment]
Added line comment.

**Comment by rubenwardy** [comment]
Fixed

**Comment by paramat** [comment]
:+1:

**Comment by paramat** [comment]
This mod needs a README.txt containing license info and texture credits (which need removing from default README).

**Comment by kwolekr** [comment]
Isn't this going to cause existing worlds to have their nyancats turn into lines of unknown nodes when the users update minetest_game?

**Comment by SmallJoker** [comment]
@kwolekr No, there are alias for this case: https://github.com/minetest/minetest_game/pull/1190/files#diff-e442119c9409de09031ed26ba9202496R79

**Comment by kwolekr** [comment]
Ahh, good, good, I missed that when looking through the changes.

I approve, +1.

File: game_api.txt [hunk 5]
```diff
@@ -250,6 +250,17 @@ Give Initial Stuff API
 ^ str is a comma separated list of initial stuff
 ^ Adds items to the list of items to be given
 
+Nyancat API
+-----------
+
+`nyancat.place(pos, facedir, length)`
+
+^ Place a cat at `pos` facing `facedir` with tail length `length`
```
Comment (kwolekr) [inline game_api.txt]: Just a nit, I feel like you should mention the restrictions on facedir here, as well as the restrictions on the Y coordinate.

File: game_api.txt [hunk 6]
```diff
@@ -250,6 +250,17 @@ Give Initial Stuff API
...
+
+`nyancat.generate(minp, maxp, seed)`
+
+^ Called by `minetest.register_on_generated`. To disable nyancat generation,
+  you can redefine nyancat.generate() to be an empty function
```
Comment (kwolekr) [inline game_api.txt]: Yuck, having to redefine functions in order to disable autogeneration...

File: mods/default/mapgen.lua [hunk 7]
```diff
@@ -1527,10 +1475,8 @@ local mg_name = minetest.get_mapgen_setting("mg_name")
 if mg_name == "v6" then
 	default.register_ores()
 	default.register_mgv6_decorations()
-	minetest.register_on_generated(default.generate_nyancats)
 elseif mg_name ~= "singlenode" then
 	default.register_biomes()
 	default.register_ores()
 	default.register_decorations()
-	minetest.register_on_generated(default.generate_nyancats)
```
Comment (kwolekr) [inline mods/default/mapgen.lua]: Now that these lines are removed, does the default game actually disable nyancat generation in these cases?

Comment (paramat) [inline mods/default/mapgen.lua]: See https://github.com/minetest/minetest_game/pull/1190#issuecomment-231827129
Because singlenode cannot be selected in world creation ```if mg_name ~= "singlenode"``` has no effect so nyancats (and other lines here) were never disabled in singlenode.
This might be a good reason to bring singlenode back to world creation?

File: game_api.txt (hunk 6 above)
Comment (paramat) [inline game_api.txt]: Yuck indeed.

File: mods/default/mapgen.lua (hunk 7 above)
Comment (paramat) [inline mods/default/mapgen.lua]: In fact these other lines are useful in singlenode to allow 'place registered ores' / 'place registered decorations' so it's best to remove ```if mg_name ~= "singlenode"``` from here.
@rubenwardy

File: game_api.txt (hunk 5 above)
Comment (rubenwardy) [inline game_api.txt]: done

**Comment by paramat** [comment]
:+1: With spelling correction.

File: game_api.txt (hunk 6 above)
Comment (kwolekr) [inline game_api.txt]: Well, I suppose it's acceptable.

**Comment by kwolekr** [comment]
:+1:

**Comment by kwolekr** [comment]
I'm convinced there exists a better approach to handling autogeneration enable/disable.
Maybe the cleanest way would be to disable it by default.  IMHO it seems downright silly to require a dependency on a mod to simply tell it that you want to disable it, if it ever happens to show up.

**Comment by paramat** [comment]
Discussion http://irc.minetest.ru/minetest-dev/2016-07-15#i_4651364

**Comment by paramat** [comment]
Actually, depending on a mod just to disable it is what already happens with the flower mod, you need to depend on it to then clear its registered decorations. So this PR seems fine to me, although i would like this to be done https://github.com/minetest/minetest_game/pull/1190#discussion_r70999779

**Comment by kwolekr** [comment]
Just because there is past precedent doesn't mean it's a good approach to take.

**Comment by paramat** [comment]
It does mean your resistence is out of proportion, since this has been a normal situation with many mods.
Because these game mods cannot be disabled they are always run whether we need their features or not. Perhaps then we can work on making game mods disableable (as has been requested) and in the meantime merge this as-is.

**Comment by paramat** [comment]
@sofar any comments?

**Comment by t4im** [comment]
Maybe https://github.com/minetest/minetest/pull/2198 should get some attention? It would solve this, too.

**Comment by paramat** [comment]
It's not ideal as that requires manual disabling of a mod. We need a way for a mod to automatically disable nyancats when used, which this PR does.

**Comment by DonBatman** [comment]
If #4307 is merged then nyan cat can just be unregistered in mods.

**Comment by paramat** [comment]
https://github.com/minetest/minetest/pull/4307
What would be unregistered?

**Comment by DonBatman** [comment]
You can unregister nyancat and rainbow block. Or even without that you can already alias them as default:stone or air. There are ways for mods to remove nyancat.

**Comment by sofar** [comment]
:+1: from me. I've always thought this was reasonable. It's never going to be pretty, but I think this is acceptable.

**Comment by t4im** [comment]
> We need a way for a mod to automatically disable nyancats when used

So why not simply:

```lua
local setting_getbool = minetest.setting_getbool
minetest.register_on_generated(function(...)
  if setting_getbool("nyancat.disable") then
    return
  end
  -- nyan nyan!
end)
```

As long as another mod has `minetest.setting_setbool("nyancat.disable", true)` in its main chunk, it will be executed before `on_genereated` without requiring a dependency.

**Comment by paly2** [comment]
:+1:
Having a way to disable nyan cat in an external mod would be really useful, and separating it into another mod of the minetest_game would make many things more simple.

**Comment by paramat** [comment]
t4im, sorry i'm not keen on that, it adds a setting and is more complex because instead of 'on-gen' containing an empty function it now looks up a setting.
I'll merge this late today with this change applied https://github.com/minetest/minetest_game/pull/1190#discussion_r70999779

**Comment by paramat** [comment]
3661cb61e37ee6b7a8818f7a28e9102fb0674e54
